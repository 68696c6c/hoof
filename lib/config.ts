class Environment {
  readonly name: string
  readonly account: number
  readonly region: string

  constructor(name: string, account: number, region: string) {
    this.name = name
    this.account = account
    this.region = region
  }
}

export class Config {
  readonly organization: string
  readonly root: Environment
  readonly prod: Environment
  readonly stage: Environment

  constructor(
    org: string,
    rootAccount: number,
    prodAccount: number,
    stageAccount: number,
    region: string,
  ) {
    this.organization = org
    this.root = new Environment('root', rootAccount, region)
    this.prod = new Environment('prod', prodAccount, region)
    this.stage = new Environment('stage', stageAccount, region)
  }
}

export default new Config('goat', 237829512642, 771299397888, 0, 'us-east-1')
