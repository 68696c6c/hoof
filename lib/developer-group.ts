import * as cdk from '@aws-cdk/core'
import * as iam from '@aws-cdk/aws-iam'

import config from './config'

export class DeveloperGroupStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props)

    const environmentName = new cdk.CfnParameter(this, 'environmentName', {
      type: 'String',
    })
    const targetAccountId = new cdk.CfnParameter(this, 'targetAccount', {
      type: 'String',
    })
    const users = new cdk.CfnParameter(this, 'users', {
      type: 'CommaDelimitedList',
    })

    const developerGroup = new iam.Group(this, 'DeveloperGroup', {
      groupName: 'DeveloperGroup',
      managedPolicies: [
        new iam.ManagedPolicy(this, 'DeveloperManagedPolicy', {
          managedPolicyName: 'DeveloperManagedPolicy',
          description: 'DeveloperGroup policy',
          document: new iam.PolicyDocument({
            statements: [
              new iam.PolicyStatement({
                actions: ['sts:AssumeRole'],
                resources: [
                  `arn:aws:iam::${targetAccountId.valueAsString}:role/${config.organization}-${environmentName.valueAsString}`,
                  `arn:aws:iam::${targetAccountId.valueAsString}:role/read-only`,
                ],
                effect: iam.Effect.ALLOW,
                conditions: {
                  Bool: {
                    'aws:MultiFactorAuthPresent': true,
                    'aws:SecureTransport': true,
                  },
                  NumericLessThan: {
                    'aws:MultiFactorAuthAge': 43200,
                  },
                },
              }),
            ],
          }),
        }),
      ],
    })

    new iam.CfnUserToGroupAddition(this, 'UserToGroupAddition', {
      groupName: developerGroup.groupName,
      users: users.valueAsList,
    })
  }
}
