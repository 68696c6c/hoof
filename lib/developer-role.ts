import * as cdk from '@aws-cdk/core'
import * as iam from '@aws-cdk/aws-iam'

import config from './config'

export class DeveloperRoleStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props)

    const environmentName = new cdk.CfnParameter(this, 'environmentName', {
      type: 'String',
    })

    new iam.Role(this, 'DeveloperRole', {
      roleName: `${config.organization}-${environmentName.valueAsString}`,
      assumedBy: new iam.PrincipalWithConditions(new iam.AccountPrincipal(config.root.account), {
        Bool: {
          'aws:MultiFactorAuthPresent': true,
          'aws:SecureTransport': true,
        },
        NumericLessThan: {
          'aws:MultiFactorAuthAge': 43200,
        },
      }),
      managedPolicies: [{ managedPolicyArn: 'arn:aws:iam::aws:policy/AdministratorAccess' }],
    })
  }
}
