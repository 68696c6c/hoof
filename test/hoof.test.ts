import { SynthUtils } from '@aws-cdk/assert'
import * as cdk from '@aws-cdk/core'

import { DeveloperRoleStack } from '../lib/developer-role'
import { DeveloperGroupStack } from '../lib/developer-group'

test('DeveloperRoleStack', () => {
  const app = new cdk.App()
  const stack = new DeveloperRoleStack(app, 'developer-role')
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
})

test('DeveloperGroupStack', () => {
  const app = new cdk.App()
  const stack = new DeveloperGroupStack(app, 'developer-group-prod')
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
})
