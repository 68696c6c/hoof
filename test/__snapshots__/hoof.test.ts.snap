// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeveloperGroupStack 1`] = `
Object {
  "Parameters": Object {
    "environmentName": Object {
      "Type": "String",
    },
    "targetAccount": Object {
      "Type": "String",
    },
    "users": Object {
      "Type": "CommaDelimitedList",
    },
  },
  "Resources": Object {
    "DeveloperGroup54FA9D31": Object {
      "Properties": Object {
        "GroupName": "DeveloperGroup",
        "ManagedPolicyArns": Array [
          Object {
            "Ref": "DeveloperManagedPolicy624C6F4D",
          },
        ],
      },
      "Type": "AWS::IAM::Group",
    },
    "DeveloperManagedPolicy624C6F4D": Object {
      "Properties": Object {
        "Description": "DeveloperGroup policy",
        "ManagedPolicyName": "DeveloperManagedPolicy",
        "Path": "/",
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Condition": Object {
                "Bool": Object {
                  "aws:MultiFactorAuthPresent": true,
                  "aws:SecureTransport": true,
                },
                "NumericLessThan": Object {
                  "aws:MultiFactorAuthAge": 43200,
                },
              },
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:aws:iam::",
                      Object {
                        "Ref": "targetAccount",
                      },
                      ":role/goat-",
                      Object {
                        "Ref": "environmentName",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:aws:iam::",
                      Object {
                        "Ref": "targetAccount",
                      },
                      ":role/read-only",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "UserToGroupAddition": Object {
      "Properties": Object {
        "GroupName": Object {
          "Ref": "DeveloperGroup54FA9D31",
        },
        "Users": Object {
          "Ref": "users",
        },
      },
      "Type": "AWS::IAM::UserToGroupAddition",
    },
  },
}
`;

exports[`DeveloperRoleStack 1`] = `
Object {
  "Parameters": Object {
    "environmentName": Object {
      "Type": "String",
    },
  },
  "Resources": Object {
    "DeveloperRole58EC5A58": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Condition": Object {
                "Bool": Object {
                  "aws:MultiFactorAuthPresent": true,
                  "aws:SecureTransport": true,
                },
                "NumericLessThan": Object {
                  "aws:MultiFactorAuthAge": 43200,
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::237829512642:root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          "arn:aws:iam::aws:policy/AdministratorAccess",
        ],
        "RoleName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "goat-",
              Object {
                "Ref": "environmentName",
              },
            ],
          ],
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
